simulation/traffic_simulator/src/entity/entity_base.cpp

```diff
@@ -713,5 +713,149 @@ auto EntityBase::updateTraveledDistance(const double step_time) -> double
+
+  if (!revert_entity_distance_to_target.has_value() && !entity_distance_to_target.has_value()) {
+    THROW_SIMULATION_ERROR("Failed to get distance between entity and target lanelet pose.");
+  }
+
+  if (!revert_entity_distance_to_target.has_value()) {
+    return entity_distance_to_target.value();
+  }
+
+  if (!entity_distance_to_target.has_value()) {
+    return revert_entity_distance_to_target.value();
+  }
+
+  if (revert_entity_distance_to_target.value() < entity_distance_to_target.value()) {
+    return revert_entity_distance_to_target.value();
+  } else {
+    return entity_distance_to_target.value();
+  }
+}
+
+bool EntityBase::reachPosition(const std::string & target_name, const double tolerance) const
+{
+  return reachPosition(other_status_.find(target_name)->second.getMapPose(), tolerance);
+}
+
+bool EntityBase::reachPosition(
+  const geometry_msgs::msg::Pose & target_pose, const double tolerance) const
+{
+  return math::geometry::getDistance(getMapPose(), target_pose) < tolerance;
+}
+
+bool EntityBase::reachPosition(
+  const CanonicalizedLaneletPose & lanelet_pose, const double tolerance) const
+{
+  return reachPosition(static_cast<geometry_msgs::msg::Pose>(lanelet_pose), tolerance);
+}
+
+auto EntityBase::requestSynchronize(
+  const CanonicalizedLaneletPose & ego_target, const CanonicalizedLaneletPose & entity_target,
+  const double threshold, const double accel_limit, const double loop_period) -> bool
+{
+  if (traffic_simulator_msgs::msg::EntityType::EGO == getEntityType().type) {
+    THROW_SYNTAX_ERROR("Request synchronize is only for non-ego entities.");
+  }
+
+  ///@brief Check if the entity has already arrived to the target lanelet.
+  if (reachPosition(entity_target, threshold)) {
+    if (this->getStatus().getTwist().linear.x < accel_limit * loop_period / 1000) {
+    } else {
+      RCLCPP_WARN_ONCE(
+        rclcpp::get_logger("traffic_simulator"),
+        "The entity has already arrived to the target lanelet and the entity is not nearly "
+        "stopped.");
+    }
+    target_speed_ = 0;
+    return true;
+  }
+
+  job_list_.append(
+    [this, ego_target, entity_target, accel_limit, threshold, loop_period](double) {
+      auto entity_distance = getDistanceToTargetLaneletPose(entity_target);
+      auto ego_distance = this->hdmap_utils_ptr_->getLongitudinalDistance(
+        static_cast<LaneletPose>(other_status_.find("ego")->second.getLaneletPose()),
+        static_cast<LaneletPose>(ego_target));
+
+      if (!ego_distance.has_value()) {
+        RCLCPP_WARN_ONCE(
+          rclcpp::get_logger("traffic_simulator"),
+          "Failed to get distance between Ego and target lanelet pose. Check if Ego has already "
+          "passed the target lanelet. If not, please contact the developer since there might be an "
+          "undiscovered bug.");
+        return true;
+      }
+      const auto ego_velocity = other_status_.find("ego")->second.getTwist().linear.x;
```

Comment:
Please check end of the iterator.
https://en.cppreference.com/w/cpp/container/unordered_map/find